pkg/cloud/cloud.go

```diff
@@ -549,40 +555,19 @@ func (c *cloud) AttachDisk(ctx context.Context, volumeID, nodeID string) (string
 
 		resp, attachErr := c.ec2.AttachVolumeWithContext(ctx, request)
 		if attachErr != nil {
-			var awsErr awserr.Error
-			if errors.As(attachErr, &awsErr) {
-				if awsErr.Code() == "VolumeInUse" {
-					return "", ErrVolumeInUse
-				}
-			}
 			return "", fmt.Errorf("could not attach volume %q to node %q: %w", volumeID, nodeID, attachErr)
 		}
 		klog.V(5).InfoS("[Debug] AttachVolume", "volumeID", volumeID, "nodeID", nodeID, "resp", resp)
 	}
 
-	attachment, err := c.WaitForAttachmentState(ctx, volumeID, volumeAttachedState, *instance.InstanceId, device.Path, device.IsAlreadyAssigned)
+	_, err = c.WaitForAttachmentState(ctx, volumeID, volumeAttachedState, *instance.InstanceId, device.Path, device.IsAlreadyAssigned)
 
 	// This is the only situation where we taint the device
 	if err != nil {
 		device.Taint()
 		return "", err
 	}
 
-	// Double check the attachment to be 100% sure we attached the correct volume at the correct mountpoint
```

Comment:
I would more describe this change as refactoring the checks, not removing them. The checks are performed within `WaitForAttachmentState`, ensuring that the attachment returned is valid/correct:

```
for _, a := range volume.Attachments {
     if a.State != nil && aws.StringValue(a.InstanceId) == expectedInstance && aws.StringValue(a.Device) == expectedDevice {
     attachmentState = aws.StringValue(a.State)
     attachment = a
   }
}
```

Reply:
The resulting code is easier to read. But if the world is in one of the unexpected states that were being explicitly tested for, we will now just silently ignore it. As with climate change, and creeping authoritarianism, I'm not sure that's a good idea.

For example, we tried to attach to `/dev/sdg` on _instanceA_, and it winds up attached to `/dev/sdh` on _instanceA_. In the old code, we'll get that clear, detailed error message from line 580. In the new code, we'll just silently time out, right? Will we even detach the miss-attached volume, so that we can retry from a known state?

Now clearly, the tests will need to be modified because multi-attach has made the world more complicated. But it seems like a regression to me, to not be checking for these cases and logging an error when they occur. Please let me know if I'm missing something.

Reply:
Extracting out the device check from this conditional and returning an error if there is a mismatch so that `AttachDisk` can be retried in case we run into an eventual consistency issue sounds like a great idea and is more resilient -- done in the latest revision.